app/modules/gpio.c

```diff
@@ -170,42 +171,95 @@ static int lgpio_write( lua_State* L )
 // gpio.mode(1,gpio.OUTPUT,gpio.PULLUP)
 // gpio.serout(1,0,{20,10,10,20,10,10,10,100}) -- sim uart one byte 0x5A at about 100kbps
 // gpio.serout(1,1,{8,18},8) -- serial 30% pwm 38k, lasts 8 cycles
+
+typedef struct
+{
+  unsigned pin;
+  unsigned level;
+  uint32 index;
+  uint32 repeats;
+  uint32 *delay_table;
+  uint32 tablelen;
+  lua_State* L;
+  int lua_done_ref; // callback when transmission is done
+} serout_t;
+static serout_t serout;
+static const os_param_t TIMER_OWNER = 0x6770696f; // "gpio"
+
+static void seroutasync_done (task_param_t arg)
+{
+  lua_rawgeti (serout.L, LUA_REGISTRYINDEX, serout.lua_done_ref);
+  lua_call (serout.L, 0, 0);
+  luaL_unref (serout.L, LUA_REGISTRYINDEX, serout.lua_done_ref);
+}
+
+static void ICACHE_RAM_ATTR seroutasync_cb(os_param_t p) {
+  (void) p;
+  NODE_DBG("%d\t%d\t%d\t%d\t%d\t%d\t%d\n", serout.repeats, serout.index, serout.level, serout.pin, serout.tablelen, serout.delay_table[serout.index], system_get_time()); // timing is delayed for short timings when debug output is enabled
+  if (serout.index < serout.tablelen) {
+    GPIO_OUTPUT_SET(GPIO_ID_PIN(pin_num[serout.pin]), serout.level);
+    serout.level = serout.level==LOW ? HIGH : LOW;
+    platform_hw_timer_arm_us(TIMER_OWNER, serout.delay_table[serout.index]);
+    serout.index++;
+    if (serout.repeats && serout.index>=serout.tablelen) {serout.index=0; serout.repeats--;}
+  } else {
+    platform_hw_timer_close(TIMER_OWNER);
+    luaM_freearray(serout.L, serout.delay_table, serout.tablelen, uint32);
+    if (serout.lua_done_ref != LUA_NOREF && serout.lua_done_ref != LUA_REFNIL) {
+      task_post_low (task_get_id((task_callback_t) seroutasync_done), (task_param_t)0);
+    }
+  }
+}
+
 static int lgpio_serout( lua_State* L )
 {
-  unsigned clocks_per_us = system_get_cpu_freq();
-  unsigned pin = luaL_checkinteger( L, 1 );
-  unsigned level = luaL_checkinteger( L, 2 );
-  unsigned repeats = luaL_optint( L, 4, 1 );
-  unsigned table_len, i, j;
+  serout.L = L;
```

Comment:
There was [a statement](https://github.com/nodemcu/nodemcu-firmware/issues/846#issuecomment-164651561) that tasks/callbacks should not re-use local states. Instead they should get it from
```lua
lua_State *L = lua_getstate();
```